services/worker-manager/src/api.js

```diff
@@ -335,3 +339,83 @@ builder.declare({
   }
   return res.reply(result);
 });
+
+builder.declare({
+  method: 'get',
+  route: '/worker/register',
+  name: 'registerWorker',
+  title: 'Register a running worker',
+  stability: APIBuilder.stability.experimental,
+  input: 'register-worker-request.yml',
+  output: 'register-worker-response.yml',
+  description: [
+    'Register a running worker.  Workers call this method on worker start-up.',
+    '',
+    'This call both marks the worker as running and returns the credentials',
+    'the worker will require to perform its work.  The worker must provide',
+    'some proof of its identity, and that proof varies by provider type.',
+  ].join('\n'),
+}, async function(req, res) {
+  const {workerPoolId, providerId, workerGroup, workerId, workerIdentityProof} = req.body;
+
+  // carefully check each value provided, since we have not yet validated the
+  // worker's "proof"
+
+  const workerPool = await this.WorkerPool.load({workerPoolId}, true);
+  if (!workerPool) {
+    return res.reportError('ResourceNotFound',
+      `Worker pool ${workerPoolId} does not exist`, {});
+  }
+
+  const provider = this.providers.get(providerId);
+  if (!provider) {
+    return res.reportError('ResourceNotFound',
+      `Provider ${providerId} does not exist`, {});
+  }
+
+  if (workerPool.providerId !== providerId && !workerPool.previousProviderIds.includes(providerId)) {
+    return res.reportError('InputError',
+      `Worker pool ${workerPoolId} not associated with provider ${providerId}`, {});
+  }
+
+  const worker = await this.Worker.load({workerPoolId, workerGroup, workerId}, true);
+  if (!worker) {
+    return res.reportError('ResourceNotFound',
+      `Worker ${workerGroup}/${workerId} in worker pool ${workerPoolId} does not exist`, {});
+  }
+
+  if (worker.providerId !== providerId) {
+    return res.reportError('InputError',
+      `Worker ${workerGroup}/${workerId} does not have provider ${providerId}`, {});
+  }
```

Comment:
Note that these are all publicly verifiable facts, so there's no reason to obscure these messages.